### ISSUE TITLE
unit test `dict-reopen` is failing

### ISSUE THREAD

**Issue by linas** [opencog/link-grammar#1088]
Running the unit test prints:
```
Opening the dictionary for the 4'th time
link-grammar: Info: Dictionary found at ../data/en/4.0.dict
link-grammar: Error: Dialect setup belongs to a different dictionary.
Fatal error: 
Assertion (NULL==sent->last_word) failed at ../../link-grammar/tokenize/tokenize.c:2806: wordgraph_create(): wordgraph exists
Segmentation fault
```

**Comment by ampli** [comment]
I have just submitted PR #1089 to address this crash.
However, an additional fix may be needed.

What happens is that the `Parse_Options` that was used to split a sentence with one dictionary, is then used to split a sentence with another dictionary. Since the dialect setting is cached in this object, it may be inappropriate for another dictionary.

The code now just notifies that without farther handling, and this is wrong (since then incorrect cached info is used). The fix here reset the said cache to enable it to be redone with the new dictionary.

But a problem remains: How to identify a new dictionary.
Possibilities:
1. Dictionary_s address.
This is what is done now. The problem with that is that a new dictionary may get the same address as the old one.
2. The proposed fix in the XXX comment - use a unique dictionary ordinal number.
The problem is that it needs to be done is a thread safe way, i.e. with tread locking, a thing that we try to avoid in the library.
3. Use a time stamp. However, it is problematic to ensure different time stamps for "concurrent" dictionary opens using portable code.
4. Use a dictionary ordinal number along with the stack frame address at the time of grabbing it.
It seems this combination must be unique.

I propose to implement possibility 4.

**Comment by linas** [comment]
Option 5: store the language in `Parse_Options` and compare it to the language in `Dictionary_s` -- if the two do not match, then there's a hard error. 

(I was thinking of creating a new release; are there pending fixes you want to get in before then?)

**Comment by ampli** [comment]
> Option 5: store the language in `Parse_Options` and compare it to the language in `Dictionary_s` -- if the two do not match, then there's a hard error.

This will not detect a the problem when the dicts happen to be named identically, e.g. if the user opens the system `en` dict and also a private `en` dict, or reopens the same dict after updating it. Since a crash may result, I think a full-proof method is needed. Option 4 just consists of incrementing a static variable and storing it along with a stack-related address. It is a few-lines solution so I think it is better.

**Comment by linas** [comment]
I don't quite understand the problem then.  Is there some cached pointer to the dict that becomes invalid when the dictionary is deleted?

**Comment by linas** [comment]
Oh, yes, I see there is. Let me look. I didn't realize that was the problem.

**Comment by ampli** [comment]
> (I was thinking of creating a new release; are there pending fixes you want to get in before then?)

I have a cleanup PR (that is almost ready) that I would like to include in the new release.
I will try to send it tomorrow.

Some other stuff  (including fake cross-links and a more aggressive pruning) is in advanced state but needs some more work so it is better not to wait for it.  (The rest is in a less advanced state.)

**Comment by linas** [comment]
Ugh... So I still don't understand (its late at night and I'm tired...) what's wrong with this pseudocode:
```
if (dinfo->cost_table != NULL) {
    if (dinfo->base_language_name != dict->name) {
          print error;
          return failure;
    }
    dinfo->dict = dict;  // just in case the dict changed
   return true;
}
```

**Comment by ampli** [comment]
This pseudocode is equivalent to the code before the last PR (#1089).
The only change in it is the less strict check of dict name instead of dict struct address.

These things are wrong in it:
1. Just printing and returning an error on a dict change without creating a new cache may result in a crash because the old cached info will be used. This is what happened before PR #1089. So it is needed to add a fix to return an error to the user (`sentence_split()` / `sentence_parse()` errors).
However, the user didn't do anything wrong at all if e.g. she/he didn't set a dialect at all and just reused the same `Parse_Options` object for another dict (this is not defined as a wrong thing to do).
2. Checking only the name doesn't detect my example cases, **so the cached info may be used for a new dict with the same name** without detecting that. In that case the result is undefined.

BTW, `dinfo->dict` is currently not used other than for detecting a dictionary change. It is a late addition, and I still need to check whether it can be used to simplify the code.

**Comment by linas** [comment]
OK, so how about this:
```
struct Dictionary_s {
   ...
   void * have_cache;
};
dictionary_create() { have_cache = NULL; }
setup_dialect() {
   if (dict->have_cache != dinfo) {...}
   dict->have_cache = dinfo;
   ...
}
```

**Comment by ampli** [comment]
This looks fine.
However, if more than one `Parse_Options` object is used per Dictionary object, the cache will need to be recalculated each time a different `Parse_Options` object is used. But since this is not expected using pattern the library, it would normally not add overhead.
 (Making `have_cache` a list would solve that but for now it seems an over complication.)

But note that the problem of using an improper dialect cache still exists in this scenario:
```
dict_a = dictionary_create_lang("a");
sent_a = sentence_create(sentence_a, dict_a);
Parse_Options po1 = parse_options_create();
sentence_split(sent_a, po1);

dict_b = dictionary_create_lang("b");
sent_b = sentence_create(sentence_b, dict_b);
parse_options_delete(po1);
Parse_Options po2 = parse_options_create(); // suppose the address of po1 is reused for po2
sentence_split(sent_b, po2);

sentence_split(sent_a, po2); // undefined behavior - may crash
```

I guess such a sequence is not likely to happen when the library is used from C.
But maybe it is more likely to happen on languages when an object (`Parse_options` in this case) may be deleted automatically.

The question is whether we would like to leave such a possibility open.

**Comment by linas** [comment]
> The question is whether we would like to leave such a possibility open.

I wouldn't bother at this time. I think multi-dictionary usage is rare; dialect usage will be rare for quite a while, looking forward. (Again, I'm interested in usig it in machine-learning settings, but this is still way off in the future)  So, mostly, the unit test needed to pass, and that's enough...

**Comment by ampli** [comment]
So I will send a PR according to your pseudocode.

**Comment by ampli** [comment]
Addressed in 7ed7fd6eedf27c437167ab089940add026859ac4.
